Bidi CCSID Support via Extension Settings

yuvalnn opened the pull request:
### Changes
This PR introduces configurable automatic CCSID conversion for non-UTF compatible source members (including BiDi CCSIDs such as 424, 420, etc.).

### Features
- New extension settings under "Source Code":
1. Automatic Conversion for non UTF compatible CCSIDs
2. Source CCSID
3. Target CCSID

- Conversion logic only runs when enabled in the settings.
- Encoding is handled via IFS-based copy to avoid corruption.
- The implementation is cleanly isolated to avoid side effects.
- Added full compatibility with "Source Dates" mode.

### How to test this PR
1. Open the extension settings → "Source Code" tab.
2. Enable Automatic Conversion for non UTF compatible CCSIDs.
3. Select a Source CCSID.
4. Select a Target CCSID.
5. Open a member encoded with the selected Source CCSID.
6. Verify the content is displayed correctly and saved as expected.
7. Open a member with a different CCSID and confirm no conversion occurs.
8. Disable conversion and confirm default behavior.

### Checklist
* [x] have tested my change
* [x] have created one or more test cases
* [ ] updated relevant documentation
* [x] Remove any/all `console.log`s I added
* [ ] have added myself to the contributors' list in [CONTRIBUTING.md](https://github.com/codefori/vscode-ibmi/blob/master/CONTRIBUTING.md)

yuvalnn committed:
Add initial Bidi support with configurable CCSID conversion

- Added two new extension settings(index.ts):
  1. Enable Bidi conversion
  2. Target CCSID for Bidi conversion

- Modified IBMiContent to apply Bidi-aware conversion logic
  when enabled via user settings, including fallback and error handling.

yuvalnn committed:
Add Bidi support to include member upload

Extends the Bidi handling to also apply when uploading source members back to the IBM i.

yuvalnn committed:
Update Bidi setting descriptions to clarify they are disabled when Source Dates are enabled

yuvalnn committed:
Merge branch 'master' into bidi-support

Itsanexpriment committed:
test(bidi): add synchronous test suite in encoding.test.ts

- Added BiDi test cases in src/api/tests/suites/encoding.test.ts
- Covers Hebrew (CCSID 424) and Arabic (CCSID 420)
- Includes helper convertToUTF8WithCCSID and edge cases

yuvalnn commented:
Hi @sebjulliand , could you please review this PR?

yuvalnn commented:
Hi @sebjulliand, thanks for the review.
That was actually my initial thought as well, but I wanted to avoid always checking the CCSID unless the user enabled it.
Thinking about it again, your approach is simpler and with no user involvement required, which seems better.
If needed, we can extend the CCSID map later.

I'll make the change and update.

Itsanexpriment committed:
feat(bidi): add source date support for bidi-compatible CCSID (#2)

* Convert to Utf8 with bidi compatible ccsid

* Convert from Utf8 to bidi ccsid when uploading member with source dates

yuvalnn committed:
Merge remote-tracking branch 'upstream/master' into bidi-support

yuvalnn committed:
feat(settings): introduce configurable CCSID conversion (source/target) in settings UI

yuvalnn committed:
fix(settings): add default option and fix CCSID selection persistence

yuvalnn committed:
Merge remote-tracking branch 'upstream/master' into bidi-support

yuvalnn committed:
Merge remote-tracking branch 'upstream/master' into bidi-support

worksofliam commented on src/api/IBMiContent.ts:
```diff
@@ -169,25 +169,57 @@ export default class IBMiContent {
+    
+    // BiDi handling if enabled
+    const isBidi = this.config.bidi;
+    const bidiCcsid = this.config.bidiCcsid;
```
Type is missing in `config`

yuvalnn committed:
test: add tests for non-BiDi source files with BiDi support enabled

- Add helper function uploadWithoutBidiDownloadWithBidi to test non-BiDi files
- Add helper function testFullCycleWithBidi to test full upload/download cycle
- Add tests for CCSID 37 and 273 with BiDi CCSID enabled
- Fix copyToImport to use stmfCcsid variable instead of hardcoded 1208

yuvalnn committed:
Merge branch 'ccsid-conversion-settings' into bidi-support

yuvalnn committed:
feat: Add CCSID validation before conversion

- Add Tools.determineCcsidConversion() to check if conversion is needed
- Check source file CCSID before applying conversion in download/upload
- Only convert if source CCSID matches configured ccsidConvertFrom
- Update tests to use new config fields
- Add CCSID conversion to extendedContent and copyToImport

yuvalnn committed:
test: enhance non-BiDi tests with edge case for mismatched CCSID

- Separate non-BiDi tests into dedicated suite for better filtering
- Add test case: file with CCSID 273 downloaded with BiDi settings 424→62211
- Verifies conversion is skipped when source CCSID doesn't match config
- Update uploadWithoutBidiDownloadWithBidi to support optional sourceCcsid parameter

yuvalnn committed:
Merge remote-tracking branch 'upstream/master' into bidi-support

yuvalnn committed:
test: restore BiDi and Non-BiDi encoding tests after merge

yuvalnn committed:
fix: run CCSID conversion commands separately to avoid job context issues

yuvalnn committed:
improve error handling in uploadMemberContent and update tests

Run SQL commands separately to avoid job context issues

yuvalnn committed:
refactor: use single await with array for SQL commands in download/upload

yuvalnn commented:
Hi @sebjulliand @worksofliam , 
Changes were made following feedback. Before performing any conversion, we now verify the source file actually matches the requested conversion criteria. if the source is not bidi, no conversion is applied, even when CCSID handling is enabled. Non-bidi sources remain unaffected.

The UI was updated to require explicit selection of both source and target CCSID.
Relevant tests were added.

sebjulliand commented on src/api/configuration/config/types.ts:
```diff
@@ -16,6 +16,9 @@ export interface ConnectionConfig extends ConnectionProfile {
   sourceDateGutter: boolean;
+  ccsidConversionEnabled: boolean;
+  ccsidConvertFrom: string;
+  ccsidConvertTo: string;
```
Default values for these three fields must be provided here (ince it's missing, it's producing an error):  https://github.com/codefori/vscode-ibmi/blob/7cc96fb211ae50a3aa77d94696ef199abb2f2546/src/api/configuration/config/ConnectionManager.ts#L47

sebjulliand commented on src/filesystems/qsys/extendedContent.ts:
```diff
@@ -172,10 +175,33 @@ export class ExtendedIBMiContent {
+        // Fetch source file CCSID and determine if conversion is needed
+        const memberPath = { library, name: file, member: name };
+        const attr = await connection.getContent().getAttributes(memberPath, "CCSID");
+        const sourceCcsid = Number(attr?.["CCSID"]) || 0;
```
See my comment about caching this.

sebjulliand reviewed (changes requested):
Thanks for your work so far @yuvalnn .
There isa conflict to resolve and some changes to be done before carrying on.

Thanks in advance!

Itsanexpriment committed:
Cache file CCSID to avoid redundant look‑ups

Itsanexpriment committed:
Change return value of determineCcsidConversion from an array to an object

Itsanexpriment committed:
Set default values to CCSID conversion fields

yuvalnn committed:
Merge upstream/master into bidi-support

- Resolved conflict in src/webviews/settings/index.ts
- Preserved SelectItem import needed for CCSID_Options
- Added new checkLoginForm import from upstream
- Synced with upstream master (commit 154bc1bd)

yuvalnn committed:
Merge helper PR: Add CCSID caching and address reviewer comments

- Add default values for CCSID config fields (ccsidConversionEnabled, ccsidConvertFrom, ccsidConvertTo)
- Change determineCcsidConversion to return object instead of array
- Implement CCSID caching in IBMi class with Map and 10-minute TTL
- Add shared getFileCcsid() method in IBMi class
- Update IBMiContent and ExtendedContent to use shared method
- Add CacheItem interface for cache entries

yuvalnn committed:
refactor(bidi-support): simplify SQL statement execution logic

yuvalnn committed:
refactor: Extract CCSID data to separate file and simplify filtering

- Move CCSID definitions to src/api/CCSIDs.ts
- Use plain objects instead of SelectItem format
- Replace Set.has() with Array.includes()
- Map to SelectItem only when needed

yuvalnn committed:
Improve CCSID cache type safety: use Map<string, CacheItem<number>>

sebjulliand reviewed (changes requested):
We're getting there @yuvalnn !
Just two changes I'd request from you to avoid duplicating code and to fix the UX in the settings.
I guess we'll be good to go afterwards. Thanks for your efforts!

yuvalnn committed:
Merge remote-tracking branch 'upstream/master' into bidi-support

yuvalnn committed:
Reuse getAttributes in getFileCcsid to avoid duplication

yuvalnn committed:
fix CCSID UI (remove placeholder + invalid targets)

sebjulliand reviewed (approved):
Good job @yuvalnn , we're good to go! 😄 
This will be released with version 3. You can check what remains to be done for v3 here if you want: https://github.com/codefori/vscode-ibmi/milestone/18
It should be done in a couple of weeks, hopefully.

In the meantime, would you mind creating a PR for that change to explain these new settings and when to use it, here in the documentation? https://github.com/codefori/docs/blob/main/src/content/docs/settings/connection.mdx

Thanks!
